lockfile_generator/src/npm.rs

```diff
@@ -33,3 +39,107 @@ impl Generator for Npm {
         "npm"
     }
 }
+
+/// Find the workspace root of an npm project.
+pub(crate) fn find_workspace_root(manifest_path: impl AsRef<Path>) -> Result<PathBuf> {
+    let manifest_path = manifest_path.as_ref();
+    let original_root = manifest_path
+        .parent()
+        .ok_or_else(|| Error::InvalidManifest(manifest_path.to_path_buf()))?
+        .canonicalize()?;
+
+    // Search parent directories for workspace manifests.
+    let mut path = original_root.as_path();
+    for _ in 0..WORKSPACE_ROOT_RECURSION_LIMIT {
+        path = match path.parent() {
+            Some(root_dir) => root_dir,
+            None => break,
+        };
+
```

Comment:
This loop could be simplified with `.ancestors()`

```suggestion
    // Search parent directories for workspace manifests.
    for path in original_root.ancestors().skip(1).take(WORKSPACE_ROOT_RECURSION_LIMIT) {
```